ml_tools/writers.py

```diff
@@ -164,6 +181,85 @@ def close(self):
         for w in self._writers:
             w.close()
 
+class WandbWriter(_MetricWriter):
+  """MetricWriter that writes to Weights & Biases (Wandb)."""
+
+  def __init__(
+    self,
+    project: str,
+    experiment: str,
+    api_key: str=None,
+    artifact_type_name: str="collection"
+  ):
+    """
+    :param project: name of the project
+    :param experiment: name of the experiment
+    :param api_key: Weights & Biases API key
+    """
+    if api_key is not None:
+      wandb.login(key=api_key)
+    logger = WandbLogger(name=experiment, project=project)
+    wandb_run_object = logger.experiment
+    assert wandb_run_object is not None
+    self._wandb_run_object = wandb_run_object
+    self._project = project
+    self._experiment = experiment
+    self._run = wandb_run_object.id
+    self._logger = logger
+    self._artifact_type_name = artifact_type_name
+
+  def log_hparams(self, hparams: Mapping[str, Any]):
+    self._logger.log_hyperparams(hparams)
+
+  def write_scalars(self, step: int, scalars):
+    self._logger.log_metrics(scalars, step=step)
+
+  def _file_collection_artifact_name(self, step: int=None):
+    project = self._project
+    experiment = self._experiment
+    run_id = self._run
+    has_step = step is not None
+    if has_step:
+      return f"{project}_{experiment}_{run_id}_{step}"
+    return f"{project}_{experiment}_{run_id}"
+
+  def write_data(self, data: Mapping[str, Union[Array, Dict]], step: int=None):
+    """
+    Writes each dictionary item to a file in directory artifact.
+    """
+    artifact_name = self._file_collection_artifact_name(step=step)
+    file_collection_artifact = wandb.Artifact(name=artifact_name, type=self._artifact_type_name)
+    for name, data_entry in data.items():
+      write_file_fn, file_ending = _resolve_write_fn(data_entry)
+      with tempfile.NamedTemporaryFile(suffix=file_ending) as fp:
+        local_filepath = fp.name
+        remote_filepath = f"{name}{file_ending}"
+        write_file_fn(filepath=local_filepath, obj=data_entry)
+        file_collection_artifact.add_file(local_path=local_filepath, name=remote_filepath)
+    wandb.run.log_artifact(file_collection_artifact) # Write artifact
+
+  def download_file_collection(self, step: int=None, target_directory: str="/tmp"):
```

Comment:
As this is a public method it might be good to add a docstring. Would it make sense for this method to return the path of the downloaded file? What happens if the file isn't downloaded?

Reply:
I'll add a docstring. Yes, that's a good idea. As it downloads a directory of files I'll make it return a list of filepaths. Found support in wandb for getting the filename. If no files are downloaded, the returned list will be empty.